tests: subsys: test cases for soc power management

s-kelvin opened the pull request:
Add a generic test case for SoC power management when supported.
Results in CPU to enter sleep/deep sleep in single/multi thread cases.
It uses hook to track entry/exit to sleep state.
It calculates sleep/deep sleep latency.

Signed-off-by: Shihao Shen <shihao.shen@intel.com>

pabigot commented:
How does this contrast with #26094 which is already in tree?  Do we need both, or can whatever's added by this instead by added to that?

mengxianglinx commented on tests/benchmarks/power_management/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  benchmark.kernel.power_management:
+    filter: CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 and CONFIG_SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1
```
This "filter" filter out all platforms, since no platform enable these config by default. Try these test cases by sanitycheck, project is compiled, but it doesn't run on any platform.

s-kelvin replied:
The test case by sanitycheck can be run on micro chip board (mec15), you can try with sanitycheck --hardware-map ~/mec1.file --device-testing -T tests/benchmarks/power_management/. This test case depends on these two configurations to run. Currently only mec1501modular_assy6885 and mec15xxevb_assy6853 have both SLEEP and DEEP_SLEEP configurations (See below)
~/zephyrproject/zephyr/boards$ grep -r SYS_PM_MIN_RESIDENCY_SLEEP_1 *
arm/mec1501modular_assy6885/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_SLEEP_1
arm/cc1352r1_launchxl/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_SLEEP_1
arm/mec15xxevb_assy6853/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_SLEEP_1
arm/cc26x2r1_launchxl/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_SLEEP_1

~/zephyrproject/zephyr/boards$ grep -r SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1 *
arm/mec1501modular_assy6885/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1
arm/mec15xxevb_assy6853/Kconfig.defconfig:config SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1

mengxianglinx replied:
got it, great!

s-kelvin commented:
> How does this contrast with #26094 which is already in tree? Do we need both, or can whatever's added by this instead be added to that?

The power management test under subsys only tests light sleep state when system enters idle threads. This one here addresses #25508 to perform operations that result in CPU to enter sleep AND deep sleep states when supported in single thread AND multithread cases, as well as some additional features such as hooks to count entry & exit, and latency measurements. I think it's hard to combine two into one since each serves a different purpose.

pabigot reviewed (changes requested):
Blocking until the following concerns are answered, if not resolved:

* Please try to make this work on boards that support SLEEP but not DEEP_SLEEP.  In particular it should build and pass on `cc1352r1_launchxl`.
* Can benchmark tests be decoupled from functional tests?
* Need to unify where the tests for this subsystem are going to go.  Unless this is primarily benchmarking it probably shouldn't be in that directory.

The test does not build on `cc1352r1_launchxl` which is one of the few accessible boards that supports power management.  The problem is it doesn't implement a deep sleep.  Can parts of this test be made conditional on support for a deep sleep mode, so that the other sleep modes can be tested?  Some SoCs have no way to support a deep sleep mode that works with the residency subsystem, and it would be unsatisfactory to not be able to test the non-deep parts.

I'm not really happy with spreading test cases all over the place.  If this is providing both functional and performance testing, can those be separated with only the performance checks here and the functional tests in `tests/subsys/power/` where all the other ones are?

At a minimum the naming should be consistent.  With this we would have:

    tests/subsys/power/power_mgmt
    tests/benchmarks/power_management

each of which is a test application, not a directory containing multiple subsystem tests.

I'd like direction from the subsystem maintainer @wentongwu and test czar @nashif.

s-kelvin commented:
> Blocking until the following concerns are answered, if not resolved:
> 
> * Please try to make this work on boards that support SLEEP but not DEEP_SLEEP.  In particular it should build and pass on `cc1352r1_launchxl`.
> * Can benchmark tests be decoupled from functional tests?
> * Need to unify where the tests for this subsystem are going to go.  Unless this is primarily benchmarking it probably shouldn't be in that directory.
> 
> The test does not build on `cc1352r1_launchxl` which is one of the few accessible boards that supports power management. The problem is it doesn't implement a deep sleep. Can parts of this test be made conditional on support for a deep sleep mode, so that the other sleep modes can be tested? Some SoCs have no way to support a deep sleep mode that works with the residency subsystem, and it would be unsatisfactory to not be able to test the non-deep parts.
> 
> I'm not really happy with spreading test cases all over the place. If this is providing both functional and performance testing, can those be separated with only the performance checks here and the functional tests in `tests/subsys/power/` where all the other ones are?
> 
> At a minimum the naming should be consistent. With this we would have:
> 
> ```
> tests/subsys/power/power_mgmt
> tests/benchmarks/power_management
> ```
> 
> each of which is a test application, not a directory containing multiple subsystem tests.
> 
> I'd like direction from the subsystem maintainer @wentongwu and test czar @nashif.

Yes I agree with the first point. I would try to make this test available for boards that support SLEEP but not DEEP SLEEP as well.
But this test is just performance testing only. Both single thread and multithread cases are serving as different scenarios for benchmarks. The rest features, such as entry/exit counter, sleep latency, and residency, are purely benchmarks.

pabigot commented:
@s-kelvin If I understand correctly you plan to update the test so it works with boards that don't have deep sleep, but that the primary purpose of this is benchmarking so it's not redundant with the other tests and doesn't need to move.

If so, please proceed.  If not, please clarify.

nashif reviewed (changes requested):
I would move this under tests/subsys/power, this does not qualify as a benchmark IMO.
wrt to supporting other SoCs with different power states, this can always be done later as long as we have the correct filters in place and as long as it is clear what is being tested.

s-kelvin commented:
> @s-kelvin If I understand correctly you plan to update the test so it works with boards that don't have deep sleep, but that the primary purpose of this is benchmarking so it's not redundant with the other tests and doesn't need to move.
> 
> If so, please proceed. If not, please clarify.

Changed the filter to SLEEP only and make sure CPU enters deep sleep only when the board has DEEP SLEEP configuration. It passed on micro chip which support both SLEEP and DEEP SLEEP also passed on boards without either SLEEP or DEEP SLEEP. However, I can't test it on cc1352r1_launchxl from my end, which supports SLEEP but not DEEP SLEEP. Please check. 
Also, I moved everything under subsys per the comments above.

nashif commented on tests/subsys/power/power_mgmt_soc/prj.conf:
```diff
@@ -0,0 +1,11 @@
+# Enable config log
```
where do you enable power management? are you assuming the boards have power management enabled?

nashif replied:
this ends up building only on the mec15xx boards, this is not a good enough coverage, those boards should not be enabling power management by default anyways, so we need another way to expand coverage where power management is supported.

s-kelvin replied:
I used filter in testcase.yaml to filter out boards that do not have sleep residency. I will add CONFIG_SYS_POWER_MANAGEMENT=y in the prj.conf to enable power management. Will this still allow mec15xx boards only?

s-kelvin replied:
Currently there are three boards that have SLEEP RESIDENCY specified and they should be covered by this test. mec15xxevb_assy6853, cc1352r1_launchxl and cc26x2r1_launchxl. I can only test the first one on my end but I don't have the other two available for testing.

s-kelvin commented:
Enabled power management and expanded coverage to mec15xx, cc1352r1_launchxl, cc26x2r1_launchxl boards.

pabigot commented:
There may need to be some parameter tuning.  cc1352r1_launchxl on this goes into deep sleep for 61 seconds each iteration, which is longer per iteration than most test infrastructures will be willing to wait for completion (over 10 minutes to finish the two 5-iteration tests).

If I set `CONFIG_SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1=1000` in `prj.conf` it completes successfully (though the deep sleep duration it prints is 2.104 s, which seems odd).

s-kelvin commented:
> There may need to be some parameter tuning. cc1352r1_launchxl on this goes into deep sleep for 61 seconds each iteration, which is longer per iteration than most test infrastructures will be willing to wait for completion (over 10 minutes to finish the two 5-iteration tests).
> 
> If I set `CONFIG_SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1=1000` in `prj.conf` it completes successfully (though the deep sleep duration it prints is 2.104 s, which seems odd).

I've set CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 to be 1000 and CONFIG_SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1 to be 2000, the same as the default residencies for mircro chip. 
The 2.104 s is because of the extra deep sleep time, DP_EXTRA_SLP_TIME=1100, as set in power_mgmt.c.

pabigot commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,326 @@
+		}
+
+		if (pm_counters[i].entry_cnt != cycles) {
+			LOG_WRN("PM counter mismatch expected: %d", cycles);
```
I get this on `cc1352r1_launchxl`:
```
I: PM sleep residency 3.104 seconds                                             
I: Resume                                                                       
ABI: PM multi-thread completed                                                  
I: PM state[0] entry counter 0                                                  
                                                                                
I: PM state[0] exit counter 0                                                   
                                                                                
W: PM counter mismatch expected: 5                                              
I: PM state[1] entry counter 0                                                  
                                                                                
I: PM state[1] exit counter 0                                                   
                                                                                
W: PM counter mismatch expected: 5                                              
Power management test finished.                                                 
 PASS - main                                                                    
===================================================================             
===================================================================             
PROJECT EXECUTION SUCCESSFUL     
```
The test passes, but there are warning-level log messages that a value isn't what is expected.

Either the test should not be passing, or those messages should be downgraded to INF.

s-kelvin replied:
cc1352 is not supposed to enter deep sleep state. I've changed the if statement at to filter out boards that support sleep only. Can you check the latest commit once again and see if cc1352 would still enter deep sleep? Thanks.

pabigot replied:
No, this isn't working.

<details>
  <summary>Click to expand trace of cc1352r1_launchxl output</summary>

```
*** Booting Zephyr OS build zephyr-v2.3.0-1348-g118693b273ce  ***
Start power management test.
I: PM single-thread test started for cycles: 3
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: PM single-thread completed
I: PM state[0] entry counter 1

I: PM state[0] exit counter 1

W: PM counter mismatch expected: 3
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 3
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 3
I: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.504 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.104 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 0

I: PM state[0] exit counter 0

W: PM counter mismatch expected: 3
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 3
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
</details>

It still says it's trying to enter deep sleep; it still enters deep sleep.

Also if those warnings indicate the test doesn't pass they need to be asserts that cause the test to fail.

s-kelvin replied:
Can you check the latest commit again and show me the output? Thanks.

pabigot replied:
Doesn't seem a lot different.  You haven't changed the warnings to assertion checks.

<details><summary>Expand for log</summary>

```
 *** Booting Zephyr OS build zephyr-v2.3.0-1348-g71ad5d8ee869  ***
Start power management test.
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: PM sleep entry latency 0.043 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.460 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM single-thread completed
I: PM state[0] entry counter 5

I: PM state[0] exit counter 5

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
</details>

s-kelvin replied:
Yea right now I want to make sure the number it enters and exits sleep/deep sleep is correct. The warning there is just comparing the number of entry/exit with total iterations, but more importantly the number of entry/exit counters should match. I will change the warnings to assertions later. For now I need to figure out why this board can't enter deep sleep but I can only see the output with the help from you.

s-kelvin replied:
Can you paste the log for the latest commit please? Thank you again.

pabigot replied:
If there's a change it isn't obvious to me.  Still says it passed, and still prints warnings about counter mismatch from expected.

<details><summary>Expand for log</summary>

*** Booting Zephyr OS build zephyr-v2.3.0-1348-g35b2cda735e7  ***
Start power management test.
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: PM sleep entry latency 0.043 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.460 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM single-thread completed
I: PM state[0] entry counter 5

I: PM state[0] exit counter 5

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
I: Thread task A init
I: Thread task B init
W: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
ABI: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
ABI: PM multi-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

W: PM counter mismatch expected: 2
I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

W: PM counter mismatch expected: 2
Power management test finished.
 PASS - main
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
</details>

s-kelvin replied:
It should be fixed with the latest commit. Only mismatches between entry counter and exit counter are evaluated now. In addition, I believe the cc1352 cannot enter deep sleep as expected because the low-level implementation for power management on the board is not correct. The test case itself is straightforward and the code logic is all correct. I don't think this should be the reason to block the merge for now. The board is supposed to support deep sleep. It enters and exits light sleep correctly but not deep sleep. But the functions of the test case are already good for both micro chips and cc1352, cc26x2.

pabigot replied:
@vanti please check this.  (I should have brought you in before, sorry, thought you were already on copy.).

vanti replied:
@pabigot Thanks for looping me in.
 
I ran the example on the cc1352, and did not see any warnings (most likely they have been fixed given the code seems to have gone through several iterations). To clarify, the cc1352 does support one deep sleep state.

Looking at the test, I see it expects deep sleep to wake up and resume from the same point of execution as where it went to sleep. That is unexpected - because to my understanding, deep sleep is not required to resume from the same point. The current definition of deep sleep in Zephyr documentation is as follows:

"CPU is power gated and loses execution context. Execution will resume at OS startup code or at a resume point determined by a bootloader that supports deep sleep resume."

On the cc1352, the deep sleep state corresponds to a shutdown state and resumes from the reset vector, hence by design the residency policy is not setup to automatically enter the deep sleep state. It wouldn't make sense for it to do so. Instead it should be manually force-triggered via sys_pm_force_power_state(). See samples/boards/ti/cc13x2_cc26x2/system_off to see how it is handled.

If we want this test to be truly generic, it should not make the assumption that the code execution would resume from the point where the CPU went into deep sleep. The way it is coded currently would make the cc1352 enter light sleep instead, which is indeed what I am seeing. I would suggest adding a comment on lines 253 and 258 of power_mgmt.c stating that platforms that do not automatically enter deep sleep states in its residency policy would simply enter a sleep state.

s-kelvin replied:
@vanti Thank you for the valuable input! I've added comments as suggested. In addition, changed to ztest framework, including adding assertions at the entry/exit counters' comparison.

s-kelvin replied:
@pabigot Can you check the latest commit? Right now it should be able to fail if there's any counter mismatch.

pabigot reviewed (changes requested):
The decreased delay makes the test pass, though it may still be getting close to one minute.    Do you need to execute the individual tests five times to be sure it's working?  3, or even 2, might be enough.

The change requested is the comment on the warning.

ceolin commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,316 @@
+/* Hooks to count entry/exit */
+void sys_pm_notify_power_state_entry(enum power_states state)
+{
+	if (!checks_enabled) {
```
why do you need  this variable / check ?

s-kelvin replied:
This is to ensure the counters are incremented only when the system goes to sleep/deep sleep caused by the two test functions. Only in the test_pwr_mgmt_singlethread and test_pwr_mgmt_multithread will the flag be assigned true so then the counters can be incremented.

ceolin commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,316 @@
+			pm_counters[i].exit_cnt);
+
+		if (pm_counters[i].entry_cnt != pm_counters[i].exit_cnt)
+			LOG_WRN("PM counters entry/exit mismatch");
```
Shouldn't the test fail here ?

s-kelvin replied:
Changed to assertion.

pabigot commented on tests/subsys/power/power_mgmt_soc/prj.conf:
```diff
@@ -0,0 +1,22 @@
+CONFIG_LOG=y
+
```
We don't normally have blank lines between config settings.

s-kelvin replied:
Done.

pabigot commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,337 @@
+		LOG_INF("PM state[%d] exit counter %d\n", i,
+			pm_counters[i].exit_cnt);
+
+		zassert_true(pm_counters[i].entry_cnt
```
This could be `zassert_equal`.

s-kelvin replied:
Done

pabigot commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,337 @@
+		pm_trigger_marker();
+		k_msleep(CONFIG_SYS_PM_MIN_RESIDENCY_SLEEP_1 +
+			 LT_EXTRA_SLP_TIME);
+		k_busy_wait(100);
```
Why are these 100 us busy waits being used after the sleeps?

s-kelvin replied:
Just put a real quick busy spinning after sleeping to avoid errors. I could remove it if it's considered unnecessary.

s-kelvin replied:
Also, I removed the userspace configuration. The LOG_PROCESS_THREAD_SLEEP_MS is set for micro chip, otherwise the latency would exceed expected values if using default. Ignore the warning for cc1352 or cc26x2.

pabigot replied:
> Just put a real quick busy spinning after sleeping to avoid errors. I could remove it if it's considered unnecessary.

Why are there errors if it's not there?  That seems like something that needs to be understood.

s-kelvin replied:
I mean I put it there just for safety. I removed it and ran it on micro chip and there were no errors.

pabigot replied:
That part was good.  Now just have to deal with this build warning from:

     rm -rf build && west build -b cc1352r1_launchxl


```
warning: LOG_PROCESS_THREAD_SLEEP_MS (defined at subsys/logging/Kconfig:200) was assigned the value
'2100' but got the value ''. Check these unsatisfied dependencies: LOG_PROCESS_THREAD (=n),
(!LOG_MINIMAL) (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_LOG_PROCESS_THREAD_SLEEP_MS.html
and/or look up LOG_PROCESS_THREAD_SLEEP_MS in the menuconfig/guiconfig interface. The Application
Development Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of
the manual might be helpful too.
```

I.e. remove that config option from `proj.conf`, or perhaps disable logging since that might affect the sleep behavior and should not affect the test.

s-kelvin replied:
Removed in the latest commit.

pabigot reviewed (changes requested):
On the code a couple minor cleanups, and a question.

There are some build warnings related to inconsistent Kconfig variables on `cc1352r1_launchxl` that you should look into.  This is the blocking issue now.

<details><summary>Build log</summary>

```
-- west build: generating a build system
Including boilerplate (Zephyr base): /mnt/nordic/zp/zephyr/cmake/app/boilerplate.cmake
-- Application: /mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc
-- Zephyr version: 2.3.99 (/mnt/nordic/zp/zephyr)
-- Found Python3: /usr/bin/python3.8 (found suitable exact version "3.8.2") found components: Interpreter 
-- Board: cc13xx
-- Aliased BOARD=cc13xx changed to cc1352r1_launchxl
-- Found toolchain: zephyr (/usr/local/zephyr-sdk)
-- Found west: /home/pab/.local/bin/west (found suitable version "0.7.3", minimum required is "0.7.1")
-- Found dtc: /usr/local/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.5.0", minimum required is "1.4.6")
-- Found BOARD.dts: /mnt/nordic/zp/zephyr/boards/arm/cc1352r1_launchxl/cc1352r1_launchxl.dts
-- Generated zephyr.dts: /mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/build/zephyr/include/generated/devicetree_unfixed.h
Parsing /mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/Kconfig
Loaded configuration '/mnt/nordic/zp/zephyr/boards/arm/cc1352r1_launchxl/cc1352r1_launchxl_defconfig'
Merged configuration '/mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/prj.conf'
Configuration saved to '/mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/build/zephyr/.config'
Kconfig header saved to '/mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/build/zephyr/include/generated/autoconf.h'

warning: USERSPACE (defined at arch/Kconfig:202) was assigned the value 'y' but got the value 'n'.
Check these unsatisfied dependencies: ARCH_HAS_USERSPACE (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_USERSPACE.html and/or look up
USERSPACE in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.


warning: LOG_PROCESS_THREAD_SLEEP_MS (defined at subsys/logging/Kconfig:200) was assigned the value
'2100' but got the value ''. Check these unsatisfied dependencies: LOG_PROCESS_THREAD (=n),
(!LOG_MINIMAL) (=n). See
http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_LOG_PROCESS_THREAD_SLEEP_MS.html
and/or look up LOG_PROCESS_THREAD_SLEEP_MS in the menuconfig/guiconfig interface. The Application
Development Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of
the manual might be helpful too.

-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/local/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc
-- Cache files will be written to: /home/pab/.cache/zephyr
-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/nordic/zp/zephyr/tests/subsys/power/power_mgmt_soc/build
-- west build: building application
[1/144] Preparing syscall dependency handling

[139/144] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       30572 B     360360 B      8.48%
      FLASH_CCFG:          88 B         88 B    100.00%
            SRAM:        7880 B        80 KB      9.62%
        IDT_LIST:         120 B         2 KB      5.86%
[144/144] Linking C executable zephyr/zephyr.elf
```
</details>

Here's the test output.
<details><summary>Latest run on cc1352r1_launchxl</summary>

```
*** Booting Zephyr OS build zephyr-v2.3.0-1348-gc704a4c1eee4  ***
Running test suite test_power_mgmt
===================================================================
START - test_pm_dummyinit
I: PM dummy single-thread test started for one cycle
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: PM sleep entry latency 0.033 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.470 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM dummy single-thread completed
 PASS - test_pm_dummyinit
===================================================================
START - test_pm_multithread
I: Thread task A init
I: Thread task B init
I: PM multi-thread test started for cycles: 2
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
I: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.102 seconds
I: Resume
I: Suspend...
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: Resume
I: Suspend...
I: About to enter deep sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: Resume
I: PM multi-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

 PASS - test_pm_multithread
===================================================================
START - test_pm_singlethread
I: PM single-thread test started for cycles: 2
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: About to enter light sleep
PM >
I: PM sleep entry latency 0.000 seconds
I: Wake from Light Sleep
PM <
I: PM sleep residency 1.503 seconds
I: About to enter deep Sleep
PM >
I: PM sleep entry latency 0.001 seconds
I: Wake from Deep Sleep
PM <
I: PM sleep residency 3.103 seconds
I: PM single-thread completed
I: PM state[0] entry counter 4

I: PM state[0] exit counter 4

I: PM state[1] entry counter 0

I: PM state[1] exit counter 0

 PASS - test_pm_singlethread
===================================================================
Test suite test_power_mgmt succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```
</details>

s-kelvin commented:
@nashif Can you please review and merge if it seems ok?

dcpleung commented on tests/subsys/power/power_mgmt_soc/src/power_mgmt.c:
```diff
@@ -0,0 +1,331 @@
+		return;
+	}
+
+	if (sys_pm_is_sleep_state(state))
+		pm_counters[0].exit_cnt++;
+	else if (sys_pm_is_deep_sleep_state(state))
+		pm_counters[1].exit_cnt++;
```
Brackets needed for if blocks, even with a single statement.

s-kelvin replied:
Thank you!. Braces added.

SecureSheII committed:
tests: subsys: test cases for soc power management

Add a generic test case for SoC power management when supported.
Results in CPU to enter sleep/deep sleep in single/multi thread cases.
It uses hook to track entry/exit to sleep state.
It calculates sleep/deep sleep latency.

Signed-off-by: Shihao Shen <shihao.shen@intel.com>

LeiW000 commented:
@nashif , it looks that kelvin has addressed your comment. Can the PR be merged now?

jocelyn-li commented:
@nashif do you have further input?
